src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -503,8 +503,32 @@ fixTsqlDefaultExpr(Archive *fout, AttrDefInfo *attrDefInfo)
 	char *runtimeErrFunc = "babelfish_runtime_error";
 	char *runtimeErrStr = "'An empty or space-only string cannot be converted into numeric/decimal data type'";
 	char *atttypname;
+	bool isBabelfishDb = isBabelfishDatabase(fout);
 
-	if (!isBabelfishDatabase(fout) ||
+	/* 
+	 * We need to re-write the decompiled DEFAULT expression for non-default
+	 * database level collation. Else, pg_dump adds explicit COLLATE clause
+	 * after the DEFAULT expression as well. This creates two explicit
+	 * COLLATE clause during dump -
+	 * 	1. For the column itself
+	 * 	2. For the default expression
+	 * Eg: CREATE TABLE t1(a nvarchar(11) DEFAULT 'default') --> 
+	 * CREATE TABLE t1("a" "sys"."nvarchar" DEFAULT 'default'::"sys"."varchar" COLLATE <non_default_collation> COLLATE <non_default_collation>)
+	 * This is an invalid syntax, hence it will fail during restore.
+	 * If, attrDefInfo->adef_expr contains COLLATE clause, it means the DEFAULT
+	 * value is collatable. Hence we re-write it to:
+	 * CREATE TABLE t1("a" "sys"."nvarchar" DEFAULT ('default'::"sys"."varchar" COLLATE <non_default_collation>) COLLATE <non_default_collation>)
+	 * by adding parenthesis. 
+	 */
+	if (isBabelfishDb && strstr(source, " COLLATE ") != NULL)
```

Comment:
Can this 'COLLATE' be in small case? In that case this comparison might fail.

Reply:
From the code I do not see that will be a problem, we always add explicit clause for Const node for default datatype like this `" COLLATE %s"`.

Reply:
Updated the logic in latest rev.